lib/whitehall_importer/migrate_assets.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module WhitehallImporter
+  class MigrateAssets
+    attr_reader :whitehall_import
+
+    def self.call(*args)
+      new(*args).call
+    end
+
+    def initialize(whitehall_import)
+      @whitehall_import = whitehall_import
+    end
+
+    def call
+      whitehall_import.migratable_assets.each do |whitehall_asset|
+        begin
+          migrate_asset(whitehall_asset)
+        rescue GdsApi::BaseError, StandardError => e
+          whitehall_asset.update!(state: "migration_failed", error_message: e.inspect)
```

Comment:
Is our decision that we will continue migrating even if assets fail? 

Something we should consider as a follow up story for this or at least for thought is what the process is if migrating assets fails. If we just carry on how does anyone know of the problem? if we don't should we abort the whole sync step.

My instinct is probably that we should run migrate assets last and let it mark the sync as failed for the whole import.

Reply:
Looking at WhitehallImporter and its calls to ResyncService and ClearLinksetLinks, it does appear that we handle all errors at the highest level, rather than in each called service - so to follow that convention, you're right that we should abort inside MigrateAssets and mark the entire sync as failed.

Would we still want to log the particular asset that it failed on? I'm thinking something like this inside MigrateAssets:

```ruby
rescue StandardError => e
  whitehall_asset.update!(state: "migration_failed", error_message: e.inspect)
  raise StandardError("Failed migrating Whitehall asset #{whitehall_asset.id}: #{e.inspect}")
end
```

Reply:
Yes we'd want to do it on a particular asset.

I actually think that in this case we might want to try do every asset and abort afterwards on the whole thing (on the logic that aborting at 1 asset failing and a remaining 10 wrong is worse than 1 out of 10 failing). It might be worth us grabbing a few mins tomorrow to chat about this as it quite nuanced and there are consistency concerns, as you rightly pointed out.

Reply:
I've had a go at implementing in the final commit 👍